pkg/blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,513 @@
+	limits *validation.Overrides,
+) (*BlockBuilder, error) {
+	b := &BlockBuilder{
+		cfg:      cfg,
+		logger:   logger,
+		register: reg,
+		limits:   limits,
+		metrics:  newBlockBuilderMetrics(reg),
+	}
+
+	instanceID, err := parseIDFromInstance(b.cfg.InstanceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse instance id: %w", err)
+	}
+
+	var ok bool
+	b.parts, ok = b.cfg.PartitionAssignment[instanceID]
+	if !ok {
+		return nil, fmt.Errorf("instance id %d must be present in partition assignment", instanceID)
+	}
+
+	bucketClient, err := bucket.NewClient(context.Background(), cfg.BlocksStorageConfig.Bucket, "block-builder", logger, reg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the bucket client: %w", err)
+	}
+	b.bucket = bucketClient
+
+	b.Service = services.NewBasicService(b.starting, b.running, b.stopping)
+
+	return b, nil
+}
+
+func parseIDFromInstance(instanceID string) (int, error) {
+	splits := strings.Split(instanceID, "-")
+	if len(splits) == 0 {
+		return 0, fmt.Errorf("malformed instance id %s", instanceID)
+	}
+
+	partID, err := strconv.Atoi(splits[len(splits)-1])
+	if err != nil {
+		return 0, fmt.Errorf("parse instance id %s, expect sequence number: %w", instanceID, err)
+	}
+
+	return partID, nil
+}
+
+func (b *BlockBuilder) starting(ctx context.Context) (err error) {
+	defer func() {
+		if err != nil {
+			if b.kafkaClient != nil {
+				b.kafkaClient.Close()
+			}
+		}
+	}()
+
+	// Empty any previous artifacts.
+	if err := os.RemoveAll(b.cfg.BlocksStorageConfig.TSDB.Dir); err != nil {
+		return fmt.Errorf("removing tsdb dir: %w", err)
+	}
+	if err := os.MkdirAll(b.cfg.BlocksStorageConfig.TSDB.Dir, os.ModePerm); err != nil {
+		return fmt.Errorf("creating tsdb dir: %w", err)
+	}
+
+	// TODO: add a test to test the case where the consumption on startup happens
+	// after the LookbackOnNoCommit with records before the after the consumption
+	// start point.
+	startAtOffsets, fallbackOffset, err := b.findOffsetsToStartAt(ctx)
+	if err != nil {
+		return fmt.Errorf("find offsets to start at: %w", err)
+	}
+	b.fallbackOffset = fallbackOffset
+
+	const fetchMaxBytes = 100_000_000
+
+	opts := []kgo.Opt{
+		kgo.ConsumePartitions(map[string]map[int32]kgo.Offset{
+			b.cfg.Kafka.Topic: startAtOffsets,
+		}),
+		kgo.FetchMinBytes(128),
+		kgo.FetchMaxBytes(fetchMaxBytes),
+		kgo.FetchMaxWait(5 * time.Second),
+		kgo.FetchMaxPartitionBytes(50_000_000),
+		// BrokerMaxReadBytes sets the maximum response size that can be read from
+		// Kafka. This is a safety measure to avoid OOMing on invalid responses.
+		// franz-go recommendation is to set it 2x FetchMaxBytes.
+		kgo.BrokerMaxReadBytes(2 * fetchMaxBytes),
+	}
+
+	metrics := ingest.NewKafkaReaderClientMetrics("block-builder", b.register)
+	opts = append(
+		commonKafkaClientOptions(b.cfg.Kafka, b.logger, metrics),
+		opts...,
+	)
+	b.kafkaClient, err = kgo.NewClient(opts...)
+	if err != nil {
+		return fmt.Errorf("creating kafka client: %w", err)
+	}
```

Comment:
Can we reuse `ingest.NewKafkaReaderClient()`? If you pick my suggestion to reuse `ingest.KafkaConfig` in block-builder then it should be easy to do.

Reply:
The hairy part is how both `ingest` and `blockbuilder` create a bootstrapping (ephemeral) admin client to construct the final Kafka client's options. And this ephemeral client needs some config too (see [1][1], [2][2]).

As mentioned above (https://github.com/grafana/mimir/pull/9068#discussion_r1732255080) I didn't want to refactor the internals of `ingest` package until you finish with the refactoring for using Kafka client from query-frontend.

[1]: https://github.com/grafana/mimir/blob/ba81e660c004f177d7c23408681d222d240790d0/pkg/storage/ingest/reader.go#L499-L506
[2]: https://github.com/grafana/mimir/blob/00f696fa4e1425e1f7c442090910ce41f2cf2da8/pkg/storage/ingest/reader.go#L428-L435